[Outreachy ] New dashboard design

RuthNjeri opened the pull request:
Fixes #8832
* [x] PR is descriptively titled 📑 and links the original issue above 🔗
* [x] tests pass -- look for a green checkbox ✔️ a few minutes after opening your PR -- or run tests locally with `rake test`
* [x] code is in uniquely-named feature branch and has no merge conflicts 📁
* [x] screenshots/GIFs are attached 📎 in case of UI updation
* [x] ask `@publiclab/reviewers` for help, in a comment below

The comments marked as TODO will be addressed either within this PR or another after some clarification.

The current issue I have is determining what query fetches the featured topics, it does not seem like they are stored in a specific table?

Some Things To Note:

The topics card display notes, wikis and possibly other items. If a topic does not have any notes, I have included a text to keep note of this. See screenshots below.

Here are the screenshots:

![Screenshot 2021-01-05 at 17 42 03](https://user-images.githubusercontent.com/7622875/103660349-8887b200-4f7e-11eb-8cc0-1a75c752afc9.png)
![Screenshot 2021-01-05 at 17 42 09](https://user-images.githubusercontent.com/7622875/103660362-8d4c6600-4f7e-11eb-8172-437795bd9520.png)
![Screenshot 2021-01-05 at 17 42 15](https://user-images.githubusercontent.com/7622875/103660367-8de4fc80-4f7e-11eb-9bad-e2536d006411.png)
![Screenshot 2021-01-05 at 17 42 28](https://user-images.githubusercontent.com/7622875/103660372-8f162980-4f7e-11eb-8524-def1758a8a02.png)
![Screenshot 2021-01-05 at 17 42 39](https://user-images.githubusercontent.com/7622875/103660374-8f162980-4f7e-11eb-8208-0088c265ee2b.png)

RuthNjeri commented:
Hi @jywarren, I couldn't find your tag name on the list of reviewers I can tag. I'm tagging you on this comment instead.
Looking forward to your feedback 😄

RuthNjeri commented on app/controllers/subscription_controller.rb:
```diff
@@ -45,7 +45,9 @@ def add
-      case tag_selection_more_than_zero?(params[:tid])
+      # The route params does not have :tid, it only accepts the name..
+      # I think the name should be used to check if the user is subscribed to a tag already
+      case tag_selection_more_than_zero?(tag.tid)
```
The route params does not have :tid, it only accepts the name.

jywarren replied:
aha, ok - i'm fine with this! Uh, are we sure, as in, can we find if this code was tested by looking at the PR it was added, and skimming the functional and system tests to check? Will we need to modify any other code if we make this change? But, totally on-board with doing this, it makes sense!

RuthNjeri replied:
I did search it in the routes folder and didn't find it, I'll have a look at the test as well to confirm.

RuthNjeri replied:
I've added a test for this, pushing it in a few. I also looked up the PR that was created for this change, no test was written and when I checked the routes folder, an add route with the params :tid was & is not present.

RuthNjeri commented on app/views/dashboard_v2/dashboard.html.erb:
```diff
@@ -0,0 +1,87 @@
+    <h4>
+      <i class="form-grey d-none d-lg-inline"><%= translation('dashboard_v2.sidebar.interest_topics')%></i>
+    </h4>
+    <%# TODO Recommended topics %>
```
I am not sure which query to use to obtain recommended topics

jywarren replied:
this should be "featured" i believe!

RuthNjeri replied:
On the mockup, it shows `recommended`.... should it be changed to featured?
![Screenshot 2021-01-06 at 20 41 17](https://user-images.githubusercontent.com/7622875/103802034-99f4bb00-505f-11eb-9f41-ee77cb08e427.png)

jywarren replied:
Yes that sounds good. I think the mockup is just older and didn't get updated to the latest wording. Thanks!!

RuthNjeri replied:
Great! Thanks!

RuthNjeri commented on app/views/dashboard_v2/dashboard.html.erb:
```diff
@@ -0,0 +1,87 @@
+<%# TODO: Topics to display if the user is new and not yet subscribed to other posts %>
```
The design showed that a new user will get an update saying "We have started you off with a few topics"
These topics should be featured. I currently don't know if there is a query I can use to show this

jywarren replied:
Hi Ruth, as of the final discussion in https://github.com/publiclab/plots2/issues/6072#issuecomment-624275876, we are not pre-selecting any topics, so we can skip that portion of the welcome message. We should just include the blog, and rely on the bottom-of-the-column featured and trending topics to encourage new people to begin subscribing. Let's see how that looks and if it's exceptionally weird we can discuss further changes! But, for now the query isn't required. 

Or did you mean a query to see if the user is new? I believe there is a conditional on the current dashboard which we can use for this.

RuthNjeri replied:
Thanks for the clarification, it was mostly about the topics to display if it's a new user. 
About the `bottom-of-the-column featured and trending topics` should I remove any topics that a user has subscribed to from that section? Let's say that the user has subscribed to a topic "test" and test is either listed in the trending or featured topics, should it be removed, or can it stay the same?

jywarren replied:
For trending topics i think we can remove already subscribed ones, but for featured, i think it'd be too tough to work with that if it's manually written... We'd have to use some brittle Javascript or something, and it's probably not worth it. 

Great!

jywarren commented on app/views/dashboard_v2/dashboard.html.erb:
```diff
@@ -0,0 +1,86 @@
+    <%# TODO Recommended topics %>
+    <br/>
+    <%# Trending topics %>
+    <small><a href="/tags" class="form-grey"><%= translation('dashboard_v2.sidebar.trending_topics') %>:</a></small>
```
perhaps... but they are 2 diff. points of interaction - one at the top when you're building your subscriptions, and one where we 'catch' people having scrolled to the bottom. Let's see how it looks and maybe trim later? We should be able to use the same feature, so not too repetitive!

jywarren replied:
oh sorry, trending vs. featured... but we can still re-use trending code!

RuthNjeri replied:
Yes! This makes sense!

jywarren commented:
Hopefully was able to answer the questions you have! 

If we can address or dismiss remaining CodeClimate issues then we can merge this and do additional work in a follow-up PR, which is probably ideal! It's getting complex enough that it's a little harder to follow, so if we can do things in smaller batches that may make reviewing smoother and we think about it iteratively. 

Thanks, @RuthNjeri you're doing a fantastic job!

jywarren commented:
Ah, and let's be sure to note the following from our discussion today in the planning issue for potentially breaking out into new issues, or at least discussing and making a decision on later:


* let's remove "empty" topics from dashboard, but ensure "follow" text at top and bottom of page links to /subscriptions.
* Another: should wiki edits or new wikis bump items to the top of the dashboard?
* Should topic cards show # of wikis?

@ebarry any thoughts on these last two?

ebarry commented:
wow impressive awesome work!

ebarry commented:
### Here's what staff talked about: 
- Should wiki edits or new wikis bump items to the top of the dashboard? **YES** When a wiki is added or edited, let’s bump up the topic card. Only display the name of the wiki in the list and not any kind of individual edit tracking
- Should topic cards show # of wikis? **YES** Let’s add number of wikis to the topic card!

### Here are other notes from our staff discussion: 
Mimi  28 minutes ago
can comments also count as activity and bump items to the top of the dashboard?

Mimi  2 minutes ago
From me: i’m wondering if there might be a way to integrate a bit more imagery on the dashboard? I love the tag pages right now and have gotten some really good feedback on how visually engaging they are. I’m excited to see the blog image and wonder if there’s a “what’s hot” or something that shows something like the cards on our current dashboard with an image for the post?

jywarren commented:
Thanks for reporting back, Liz! Regarding all of the staff feedback, let's be sure to move it all to its own issue and think on it in a later cycle of development, so as not to let it interrupt progress in this PR, sound good @RuthNjeri?

RuthNjeri commented:
Thanks for the updates @ebarry.
Yes, I agree @jywarren, I can list all the TODOs and what Liz has mentions as separate issues. Also, I have addressed all of the feedback and added one more test for the `add` route. I think it will be ready to merge once all the tests have passed.

sagarpreet-chadha reviewed (commented):
Great work so far !!!

Tlazypanda reviewed (commented):
@RuthNjeri This looks amazing :tada: :100:

jywarren commented:
Looks perfect. Maybe later we may want to move the sidebar for the dashboard back into the dashboard folder? I don't really know if the sidebar views folder is really helping organization... but open to your suggestions!

Merging! Let's see this boot up on stable.publiclab.org!!!!

jywarren commented:
Congrats!!!!!! 🎉 🎉 🎉 🎉 

thanks, everyone!!!

cesswairimu commented:
Great job @RuthNjeri :tada: :tada: 

Looks like stable is broken :cry:

RuthNjeri commented:
> Looks perfect. Maybe later we may want to move the sidebar for the dashboard back into the dashboard folder? I don't really know if the sidebar views folder is really helping organization... but open to your suggestions!
> 
> Merging! Let's see this boot up on stable.publiclab.org!!!!

@jywarren I totally agree. I was of the same opinion as well.. I'll add this as a task/issue
